reasoner/resolution/resolver/BoundConcludableResolver.java

```diff
@@ -0,0 +1,415 @@
+            cache = new ConceptMapCache(new HashMap<>(), bounds, () -> traversalIterator(concludable.pattern(), bounds));
+            boolean singleAnswerRequired = bounds.concepts().keySet().containsAll(unboundVars);
+            requestState = new Match(fromUpstream, cache.asConceptMapCache(), iteration, singleAnswerRequired);
+            requestState.asExploration().downstreamManager().addDownstreams(ruleDownstreams(fromUpstream));
+        }
+        return requestState;
+    }
+
+    private List<Request> ruleDownstreams(Request fromUpstream) {
+        // loop termination: when receiving a new request, we check if we have seen it before from this root query
+        // if we have, we do not allow rules to be registered as possible downstreams
+        List<Request> downstreams = new ArrayList<>();
+        AnswerState.Partial.Concludable<?> partialAnswer = fromUpstream.partialAnswer().asConcludable();
+        for (Map.Entry<Driver<ConclusionResolver>, Set<Unifier>> entry : applicableRules.entrySet()) {
+            Driver<ConclusionResolver> conclusionResolver = entry.getKey();
+            for (Unifier unifier : entry.getValue()) {
+                Optional<? extends AnswerState.Partial.Conclusion<?, ?>> unified = partialAnswer.toDownstream(
+                        unifier, resolverRules.get(conclusionResolver));
+                unified.ifPresent(
+                        conclusion -> downstreams.add(Request.create(driver(), conclusionResolver, conclusion)));
+            }
+        }
+        return downstreams;
+    }
+
+    private class FollowingMatch extends RequestState.CachingRequestState<ConceptMap, ConceptMap> {
+
+        public FollowingMatch(Request fromUpstream, AnswerCache<ConceptMap, ConceptMap> answerCache, int iteration) {
+            super(fromUpstream, answerCache, iteration, true, true);
+        }
+
+        @Override
+        protected FunctionalIterator<? extends AnswerState.Partial<?>> toUpstream(ConceptMap conceptMap) {
+            return Iterators.single(fromUpstream.partialAnswer().asConcludable().asMatch().toUpstreamLookup(
+                    conceptMap, concludable.isInferredAnswer(conceptMap)));
+        }
+
+    }
+
+    private class Match extends RequestState.CachingRequestState<ConceptMap, ConceptMap> implements RequestState.Exploration {
+
+        private final DownstreamManager downstreamManager;
+        private final boolean singleAnswerRequired;
+
+        public Match(Request fromUpstream, AnswerCache<ConceptMap, ConceptMap> answerCache,
+                     int iteration, boolean singleAnswerRequired) {
+            super(fromUpstream, answerCache, iteration, true, false);
+            this.downstreamManager = new DownstreamManager();
+            this.singleAnswerRequired = singleAnswerRequired;
+        }
+
+        @Override
+        public boolean isExploration() {
+            return true;
+        }
+
+        @Override
+        public Exploration asExploration() {
```

Comment:
fyi  i think it looked like `Exploration` is only used in `BoundConcludable` -- do we want to move it here?

Reply:
If we did that then surely the casting won't work?